atef/widgets/config/find_replace.py

```diff
@@ -0,0 +1,1009 @@
+        else:
+            logger.warning('Unable to generate preview, provided replacement '
+                           'text is invalid')
+            post_text = '[INVALID]'
+        self.post_label.setText(f'<font>{post_text}</font>')
+        self.post_label.setToolTip(f'<font>{post_text}</font>')
+
+        ok_button = self.button_box.button(QtWidgets.QDialogButtonBox.Ok)
+        ok_button.setText('')
+        ok_button.setIcon(qta.icon('ei.check'))
+        delete_button = self.button_box.button(QtWidgets.QDialogButtonBox.Cancel)
+        delete_button.setText('')
+        delete_icon = self.style().standardIcon(
+            QtWidgets.QStyle.SP_TitleBarCloseButton
+        )
+        delete_button.setIcon(delete_icon)
+
+        # path details
+        path_list = []
+        for segment in self.data.path:
+            if isinstance(segment[0], str):
+                name = segment[0]
+            else:
+                name = type(segment[0]).__name__
+            path_list.append(f'({name}, {segment[1]})')
+
+        path_str = '>'.join(path_list)
+        detail_widget = QtWidgets.QLabel(path_str + '\n')
+        detail_widget.setWordWrap(True)
+
+        widget_action = QtWidgets.QWidgetAction(self.details_button)
+        widget_action.setDefaultWidget(detail_widget)
+
+        widget_menu = QtWidgets.QMenu(self.details_button)
+        widget_menu.addAction(widget_action)
+        self.details_button.setMenu(widget_menu)
+
+        self.button_box.accepted.connect(self.apply_action)
+        self.button_box.rejected.connect(self.reject_action)
+
+    def apply_action(self) -> None:
+        success = self.data.apply()
+        if not success:
+            QtWidgets.QMessageBox.warning(
+                self,
+                'Edit was not applied successfully, and will be removed'
+            )
+        self.remove_item.emit()
+
+    def reject_action(self) -> None:
+        self.remove_item.emit()
+
+
+class FillTemplatePage(DesignerDisplay, QtWidgets.QWidget):
+
+    file_name_label: QtWidgets.QLabel
+    type_label: QtWidgets.QLabel
+
+    device_table: QtWidgets.QTableWidget
+    # TODO?: filter by device type?  look at specific device types?
+    vert_splitter: QtWidgets.QSplitter
+    horiz_splitter: QtWidgets.QSplitter
+    details_list: QtWidgets.QListWidget
+    edits_table: TableWidgetWithAddRow
+    edits_table_placeholder: QtWidgets.QWidget
+    # TODO?: smart initialization?  Choosing edits by clicking on devices?
+    # TODO?: starting device / string, happi selector for replace?
+
+    apply_all_button: QtWidgets.QPushButton
+    open_button: QtWidgets.QPushButton
+    save_button: QtWidgets.QPushButton
+    verify_button: QtWidgets.QPushButton
+
+    busy_thread: Optional[BusyCursorThread]
+
+    filename = 'fill_template_page.ui'
+
+    def __init__(
+        self,
+        *args,
+        filepath: Optional[str] = None,
+        window: Optional[Any] = None,
```

Comment:
Probably can do better than `Any` here

Reply:
TIL `typing.TYPE_CHECKING` was a thing.  circular imports be gone!